x-pack/spec/filters/geoip/database_manager_spec.rb

```diff
@@ -13,215 +13,249 @@
-    end
-
-    context "patch database" do
-      it "use input path" do
-        path = db_manager.send(:patch_database_path, default_asn_db_path)
-        expect(path).to eq(default_asn_db_path)
-      end
-
-      it "use CC license database as default" do
-        path = db_manager.send(:patch_database_path, "")
-        expect(path).to eq(default_city_db_path)
-      end
+    CITY = LogStash::Filters::Geoip::CITY
+    ASN = LogStash::Filters::Geoip::ASN
+    CC = LogStash::Filters::Geoip::CC
 
-      it "failed when default database is missing" do
-        expect(db_manager).to receive(:file_exist?).and_return(false, false)
-        expect { db_manager.send(:patch_database_path, "") }.to raise_error /I looked for/
-      end
+    before do
+      stub_const('LogStash::Filters::Geoip::DownloadManager::GEOIP_ENDPOINT', "https://somewhere.dev")
+      allow(mock_geoip_plugin).to receive(:update_filter)
     end
 
-    context "md5" do
-      it "return md5 if file exists" do
-        str = db_manager.send(:md5, default_city_db_path)
-        expect(str).not_to eq("")
-        expect(str).not_to be_nil
-      end
-
-      it "return empty str if file not exists" do
-        file = Stud::Temporary.file.path + "/invalid"
-        str = db_manager.send(:md5, file)
-        expect(str).to eq("")
-      end
+    after do
+      LogStash::Filters::Geoip::DatabaseManager.class_variable_set(:@@instance, nil)
+      delete_file(metadata_path)
     end
 
-    context "check age" do
-      it "should raise error when 30 days has passed" do
-        expect(mock_metadata).to receive(:updated_at).and_return((Time.now - (60 * 60 * 24 * 33)).to_i)
-        expect(mock_metadata).to receive(:cc?).and_return(false)
-        expect{ db_manager.send(:check_age) }.to raise_error /be compliant/
+    context "initialize" do
+      it "should set the initial state to cc database" do
+        states = db_manager.instance_variable_get(:@states)
+        expect(states[CITY].is_eula).to be_falsey
+        expect(states[CITY].database_path).to eql(states[CITY].cc_database_path)
+        expect(::File.exist?(states[CITY].cc_database_path)).to be_truthy
+        expect(states[ASN].is_eula).to be_falsey
+        expect(states[ASN].database_path).to eql(states[ASN].cc_database_path)
+        expect(::File.exist?(states[ASN].cc_database_path)).to be_truthy
       end
 
-      it "should give warning after 25 days" do
-        expect(mock_metadata).to receive(:updated_at).and_return((Time.now - (60 * 60 * 24 * 26)).to_i)
-        expect(mock_metadata).to receive(:cc?).and_return(false)
-        expect(mock_geoip_plugin).to receive(:terminate_filter).never
-
-        if db_manager.instance_variable_get(:@mode) != :disabled
-          expect(LogStash::Filters::Geoip::DatabaseManager).to receive(:logger).at_least(:once).and_return(logger)
-          expect(logger).to receive(:warn)
-          expect(logger).to receive(:info)
+      context "when metadata exists" do
```

Comment:
Regarding the DatabaseManager instance, we can create a throwaway class for the test. This also simplifies testing, where we separate the testing subject - the database manager, from the setup steps (creating metadata).

I'd also suggest adding a simple accessor to the DatabaseManager to make it easier to retrieve states, thus coming to this:

```ruby
        let(:db_manager) { Class.new(LogStash::Filters::Geoip::DatabaseManager).instance }

        before do
          FileUtils.cp_r(get_dir_path(CC), get_dir_path(second_dirname))
          write_temp_metadata(metadata_path, city2_metadata)
        end

        it "should use database record in metadata" do
          states = db_manager.state(CITY)
          expect(state.is_eula).to be_truthy
          expect(state.database_path).to include second_dirname
        end
```

We could even move the `states = db_manager.state(CITY)` to a let so that the `it` block only consists of these two expectations, but I'm ok with this too.

Reply:
The throwaway class helps a lot :) thanks! 

I didn't expose `state` because the one which takes the state could update this internal value

Reply:
This is a very valid concern, which can be addressed by duping the hash before returning it.